ignitionrobotics/ign-gazebo
# Add comms infrastructure

# 🎉 New feature

Requires

* https://github.com/ignitionrobotics/ign-msgs/pull/238

## Summary


This PR adds some infrastructure for being able to create comms systems. A comms system lets you send and receive data under the constrains of a given comms model. The comms model specifies when and how messages should be delivered to its destinations.

Subscribers create a regular callback via Ignition Transport and send a request to a centralized broker to bind an address, the robot attached to the address and the topic name used as a callback. Publishers fill a `Dataframe` message specifying the addresses of the sender and destination and set the payload. Then, publishers need to send via Ignition Transport the message to the centralized broker.

Besides the general infrastructure, this PR contains two systems: `PerfectComms` and `CommsEndpoint`.

`PerfectComms` is an example of a comms system. As required, it implements the `ICommsModel` interface and it always delivers the messages to their destinations.

`CommsEndpoint` is a helper system that can be optionally attached to a model (see `examples/worlds/comms.sdf`) and lets you assign an address to a robot and the subscription topic. The system automatically connects with the broker and bind this address/robot for you. You could do this process manually if needed but I think it's convenient for most of the users.

## Test it


* Run the tests.
* Run the example:

1. Launch Gazebo:
```
ign gazebo -v 4 <your_ws>/src/ign-gazebo/examples/worlds/perfect_comms.sdf
```
2. Launch a subscriber:
```
ign topic -e -t addr2/rx
```
3. Launch a publisher:
```
cd <your_ws>/src/ign-gazebo/examples/standalone/comms
mkdir build && cd build
cmake ..
make
./publisher addr2
```
You should observe messages on the subscriber terminal.

CC @arjo129 

## Checklist

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸
